tools/common/Driver.cs

```diff
@@ -235,54 +272,110 @@ public static void Log (int min_verbosity, string format, params object [] args)
 					throw ErrorHelper.CreateError (68, Errors.MX0068, fx);
-#if MONOMAC
-				if (parsedFramework == TargetFramework.Net_3_0 || parsedFramework == TargetFramework.Net_3_5)
-					parsedFramework = TargetFramework.Net_2_0;
-#endif
 
 				targetFramework = parsedFramework;
 
 				break;
 			}
 
-#if MTOUCH
-			if (Array.IndexOf (TargetFramework.ValidFrameworks, targetFramework.Value) == -1)
-				throw ErrorHelper.CreateError (70, Errors.MT0070, targetFramework.Value, string.Join (" ", TargetFramework.ValidFrameworks.Select ((v) => v.ToString ()).ToArray ()));
+			bool show_0090 = false;
+#if MONOMAC
+			if (!TargetFramework.IsValidFramework (targetFramework)) {
+				// For historic reasons this is messy.
+				// If the TargetFramework we got isn't any of the one we accept, we have to do some fudging.
+				bool force45From40UnifiedSystemFull = false;
+
+				if (references.Any ((v) => Path.GetFileName (v) == "XamMac.dll"))
+					throw ErrorHelper.CreateError (143, Errors.MM0143);
+
+				if (targetFramework == TargetFramework.Net_2_0
+					|| targetFramework == TargetFramework.Net_3_0
+					|| targetFramework == TargetFramework.Net_3_5
+					|| targetFramework == TargetFramework.Net_4_0
+					|| targetFramework == TargetFramework.Net_4_5) {
+					// .NETFramework,v2.0 => Xamarin.Mac,Version=v4.5,Profile=Full
+					// .NETFramework,v3.0 => Xamarin.Mac,Version=v4.5,Profile=Full
+					// .NETFramework,v3.5 => Xamarin.Mac,Version=v4.5,Profile=Full
+					// .NETFramework,v4.0 => Xamarin.Mac,Version=v4.5,Profile=Full
+					// .NETFramework,v4.5 => Xamarin.Mac,Version=v4.5,Profile=Full
+					TargetFramework = TargetFramework.Xamarin_Mac_4_5_Full;
+				} else if (TargetFramework.Identifier == TargetFramework.Xamarin_Mac_2_0_Mobile.Identifier
+					&& TargetFramework.Version == TargetFramework.Xamarin_Mac_2_0_Mobile.Version) {
+					// At least once instance of a TargetFramework of Xamarin.Mac,v2.0,(null) was found already. Assume any v2.0 implies a desire for Modern.
+					TargetFramework = TargetFramework.Xamarin_Mac_2_0_Mobile;
+				} else if (TargetFramework.Identifier == TargetFramework.Xamarin_Mac_4_5_Full.Identifier
+						 && TargetFramework.Profile == TargetFramework.Xamarin_Mac_4_5_Full.Profile) {
+					// Xamarin.Mac,Version=vX.Y,Profile=Full => Xamarin.Mac,Version=v4.5,Profile=Full
+					TargetFramework = TargetFramework.Xamarin_Mac_4_5_Full;
+				} else if (TargetFramework.Identifier == TargetFramework.Xamarin_Mac_4_5_System.Identifier
+						 && TargetFramework.Profile == TargetFramework.Xamarin_Mac_4_5_System.Profile) {
+					// Xamarin.Mac,Version=vX.Y,Profile=System => Xamarin.Mac,Version=v4.5,Profile=System
+					TargetFramework = TargetFramework.Xamarin_Mac_4_5_System;
+				} else {
+					// This is a total hack. Instead of passing in an argument, we walk the references looking for
+					// the "right" Xamarin.Mac and assume you are doing something
+					// Skip it if xamarin-full-framework or xamarin-system-framework passed in 
```

Comment:
did not the commit message mention those `xamarin-[full|system]-framework` were removed ?

Reply:
They're deprecated, not removed, so they'll still work. But while the comment is still technically correct, it's still outdated, so I'll update it.